File: src/main/java/fr/exanpe/t5/lib/components/HideablePanel.java
//
// Copyright 2011 EXANPE <exanpe@gmail.com>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package fr.exanpe.t5.lib.components;

import org.apache.tapestry5.Asset;
import org.apache.tapestry5.BindingConstants;
import org.apache.tapestry5.Block;
import org.apache.tapestry5.ClientElement;
import org.apache.tapestry5.ComponentResources;
import org.apache.tapestry5.MarkupWriter;
import org.apache.tapestry5.annotations.AfterRender;
import org.apache.tapestry5.annotations.BeginRender;
import org.apache.tapestry5.annotations.Import;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.annotations.Property;
import org.apache.tapestry5.annotations.SupportsInformalParameters;
import org.apache.tapestry5.dom.Element;
import org.apache.tapestry5.ioc.annotations.Inject;
import org.apache.tapestry5.json.JSONObject;
import org.apache.tapestry5.services.javascript.JavaScriptSupport;

import fr.exanpe.t5.lib.services.ExanpeComponentService;

/**
 * This component renders a left panel to be shown/hidden on will.<br/>
 * JavaScript : This component is bound to a class Exanpe.HideablePanel<br/>
 * CSS : This component is bound to a class exanpe-hdp.<br/>
 * 
 * @author Jul
 */
@Import(library =
{ "js/exanpe-t5-lib.js", "${exanpe.yui2-base}/yahoo-dom-event/yahoo-dom-event.js", "${exanpe.yui2-base}/animation/animation-min.js" }, stylesheet =
{ "css/exanpe-t5-lib-core.css", "css/exanpe-t5-lib-skin.css" })
@SupportsInformalParameters
public class HideablePanel implements ClientElement
{
    /**
     * Bloc hideable, disposed on left
     */
    @Property
    @Parameter(required = true, defaultPrefix = BindingConstants.LITERAL)
    @SuppressWarnings("unused")
    private Block hideable;

    /**
     * Bloc always visible, on right
     */
    @Property
    @Parameter(required = true, defaultPrefix = BindingConstants.LITERAL)
    @SuppressWarnings("unused")
    private Block visible;

    /**
     * Width in pixel for the hideable panel
     */
    @Property
    @Parameter(required = true, defaultPrefix = BindingConstants.LITERAL)
    private int hideWidth;

    /**
     * Duration of show
     */
    @Property
    @Parameter(required = true, defaultPrefix = BindingConstants.LITERAL, value = "1.0")
    private float duration;

    /**
     * Image displayed to hide the panel
     */

    @Property
    @Parameter(defaultPrefix = BindingConstants.ASSET, value = "img/hdp/hide.png")
    @SuppressWarnings("unused")
    private Asset hideImg;

    /**
     * Image displayed to show the panel
     */
    @Property
    @Parameter(defaultPrefix = BindingConstants.ASSET, value = "img/hdp/show.png")
    @SuppressWarnings("unused")
    private Asset showImg;

    private static final String ROOT_CSS_CLASS = "exanpe-hdp";

    @Inject
    private ComponentResources resources;

    @Inject
    private JavaScriptSupport javascriptSupport;

    @Inject
    private ExanpeComponentService ecservice;

    @BeginRender
    void begin(MarkupWriter writer)
    {
        Element e = writer.element("table");

        e.attribute("id", getClientId());
        resources.renderInformalParameters(writer);

        e.attribute("cellspacing", "0");
        e.attribute("cellpadding", "0");

        ecservice.reorderCSSClassDeclaration(e, ROOT_CSS_CLASS);
    }

    @AfterRender
    void end(MarkupWriter writer)
    {
        writer.end();

        JSONObject data = buildJSONData();

        javascriptSupport.addInitializerCall("hideablePanelBuilder", data);
    }

    private JSONObject buildJSONData()
    {
        JSONObject data = new JSONObject();

        data.accumulate("id", getClientId());
        data.accumulate("hideWidth", hideWidth);
        data.accumulate("duration", this.duration);

        return data;
    }

    public String getClientId()
    {
        return resources.getId();
    }
}

File: src/main/resources/META-INF/modules/exanpe/hideablePanel.js
/*
 * Copyright 2013 EXANPE <exanpe@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

define(["exanpe/common",
		"yui/2.9.0/yahoo-dom-event", 
		"yui/2.9.0/animation-min"], function(hideablePanel) {

	/** 
	 * Constructor
	 * @class Represents an Hideable Panel. Can be accessed through JavaScript by its id.
	 * @param {String} id the id of the main accordion element
	 * @param {int} hideWidth the width of the hideable panel
	 * @param {float} duration the speed of hidding animation
	 */
	Exanpe.HideablePanel = function(id, hideWidth, duration){
		/**
		 * id of the instance
		 */
		this.id = id;
		
		/**
		 * duration of the animation
		 */
		this.duration = duration;
		
		/**
		 * state of the panel
		 */
		this.hidden = false;
		
		var attributesHide = { 
	        width: { to: 0 }
	    }; 
		this.animHide = new YAHOO.util.Anim(this.id+'_hidepart', attributesHide, this.duration);
		this.animHide.onComplete.subscribe( function(){
			YAHOO.util.Dom.addClass(id, 'hidden');
		});
		
		var attributesShow = { 
	        width: { to: hideWidth  ,unit: 'px'}
	    }; 
		this.animShow = new YAHOO.util.Anim(this.id+'_hidepart', attributesShow, this.duration); 
		this.animShow.onComplete.subscribe( function(){
			YAHOO.util.Dom.removeClass(id, 'hidden');
		});
		
		var hideBar = this.getHideBarEl();
		
		YAHOO.util.Event.addListener(hideBar, "click", Exanpe.HideablePanel.prototype.reverse, null, this);
	    YAHOO.util.Event.addListener(hideBar, "mouseover", function(ev, id){YAHOO.util.Dom.addClass(hideBar, 'hidebarHover');}, this.id);
	    YAHOO.util.Event.addListener(hideBar, "mouseout", function(ev, id){YAHOO.util.Dom.removeClass(hideBar, 'hidebarHover');}, this.id);
	};

	/**
	 * Method to get the DOM element of the hide bar
	 * @return {HTMLElement} the hidding bar
	 * @private 
	 */
	Exanpe.HideablePanel.prototype.getHideBarEl = function(){
		return YAHOO.util.Dom.get(this.id+"_hidebar");
	};

	/**
	 * Reverse the visibility of the panel
	 * @return {boolean} true if the panel gets visible, false if it gets hidden
	 */
	Exanpe.HideablePanel.prototype.reverse = function(){
		if(this.hidden){
			this.hidden = false;
			this.animShow.animate();
			this.afterShow();
			return true;
		}else{
			this.hidden = true;
			this.animHide.animate();
			this.afterHide();
			return false;
		}	
	};

	/**
	 * Returns the visibility of the panel
	 * @return {boolean} true if the panel is visible, false otherwise
	 */
	Exanpe.HideablePanel.prototype.isVisible = function(){
		return !this.hidden;
	};

	/**
	 * Called after hide.
	 * Does nothing by default, override to define your own action.
	 * @event
	 */
	Exanpe.HideablePanel.prototype.afterHide = function(){
		
	};

	/**
	 * Called after show.
	 * Does nothing by default, override to define your own action.
	 * @event
	 */
	Exanpe.HideablePanel.prototype.afterShow = function(){
		
	};


	/**
	 * Initializes an hideable panel on dom load
	 * @param {Object} data the json data coming from Java class initialization
	 * @private
	 * @static
	 */
	function hideablePanelBuilder(data) {
		var hdp = new Exanpe.HideablePanel(data.id, data.hideWidth, data.duration);
		window[hdp.id] = hdp;
	};

	return {
		init: hideablePanelBuilder
	}
});